### DISCUSSION TITLE
Edit Routing for Transforms

### DISCUSSION THREAD

**Discussion by mwalk176** [Autodesk/maya-usd#3254] [Q&A]
Hi everyone, I have a composed stage of multiple usd files and am trying to get Edit Routing working so each time an artist changes a prim, the edits end up in the proper sublayer. I have visibility and attributes working, but it seems like when you translate, rotate, or scale the prim, those changes aren't considered attributes and do not get picked up by the EditRouter. Has that functionality been implemented yet? Or would those operations be able to be done under say, a composite routing command?

**Comment by santosd** [comment]
Hi @mwalk176 I was speaking with @pierrebai-adsk who did a lot of the work for edit routing and he let me know that it seems that the Maya manipulators by pass edit routing so only the attributes editing in the AE, channel box, UFE attributes or through setAttr command are routed. Though it seems that the manipulators are creating UFE commands so it should be possible to route through those commands. I hope that helps.

**Reply by mwalk176** [reply]
That makes sense, thank you @santosd! I haven't had much experience with UFE Commands yet, do you know how I can access those created UFE commands from the manipulators in Python? Or would I have to modify the maya-usd plugin in C++ to access those?

**Reply by santosd** [reply]
@mwalk176 hmm... based off of the information I was able to find in the documentation: https://github.com/Autodesk/maya-usd/blob/dev/doc/EditRouting.md#maya-references there doesn't seem to be `move` or `xform` in the subset of the commands that can be routed, but maybe I am wrong. @pierrebai-adsk are you familiar with how this could be done?

**Comment by pierrebai-adsk** [comment]
You are right that not all commands are currently routable. For your particular problem, the commands used to change the transforms on prims have not been made routable yet. While the code changes required to make one command routable is not extensive, there are a lot of commands for the manipulators that would all need to be modified simultaneously.

**Reply by mwalk176** [reply]
Thanks for the update @pierrebai-adsk, if I was to potentially implement this functionality myself, I assume I could look at how it was done for previous commands like visibility (https://github.com/Autodesk/maya-usd/pull/2546/files) for reference, but what commands do the manipulators specifically call? Is that all UFE?

**Comment by pierrebai-adsk** [comment]
Yes, you would have to mimick how it was done for the visibility command.

For the Maya manipulators, the way it works is that Maya ask the UFE run-time manager to create a Ufe::Transform3d object for a given prim. These Ufe::Transform3d have functions to create UFE command corresponding to the 3D manipulation. In the MayaUsd plugin, we have several implementations of the Ufe::Transform3d interface to support different USD "transform-stack", that is several ways to describe transformation: the Maya one, the USD common one, fallbacks, etc.

A while ago I wrote a presentation about the way the Transform3d is implemented in the plugin, which you can find [here](https://github.com/Autodesk/maya-usd/blob/dev/doc/UsdTransformsStack.pdf)

An example of one of the implementation of the Transform3d interface can be found [UsdTransform3dMayaXformStack.h](https://github.com/Autodesk/maya-usd/blob/dev/lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.h). You would need to find all classes that derive from Ufe::Transform3d. So you would need to implement edit routing for *all* of these commands for all these implementations, which might be a few dozen commands.

**Reply by mwalk176** [reply]
I see, thank you for the resources! And that's a good heads up, I'll see what I can do about those commands.

**Comment by santosd** [comment]
@mwalk176 when you get this implemented for manipulators let us know so we can take a look and hopefully merge this into the main branch. That would be awesome!

**Reply by mwalk176** [reply]
I'll keep you posted! Thanks @santosd!
